File: src/main/java/com/zippeykeys/praisebe/registry/PBRegistry.java
package com.zippeykeys.praisebe.registry;

import java.lang.reflect.InvocationTargetException;
import java.util.Arrays;
import java.util.EnumMap;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Set;
import java.util.function.BiConsumer;

import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableSet;
import com.zippeykeys.praisebe.deity.Deity;
import com.zippeykeys.praisebe.util.ClassUtil;

import org.immutables.builder.Builder.Factory;
import org.immutables.value.Value.Style;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.Nullable;

import lombok.val;
import lombok.var;
import net.minecraft.util.ResourceLocation;

@Style(depluralize = true, strictBuilder = true)

public class PBRegistry<T> {
    protected final ImmutableSet<Class<? extends Enum<?>>> classifiers;

    protected final Map<String, T> classes;

    protected final Map<Class<? extends Enum<?>>, Map<Enum<?>, Set<T>>> categorized;

    @Factory
    @Contract("_, _ -> new")
    @SuppressWarnings("unchecked")
    public static <T> PBRegistry<T> pBRegistry(Class<T> clazzT, List<Class<? extends Enum<?>>> classifiers) {
        if (clazzT == Deity.class)
            return (PBRegistry<T>) new Deity.Registry();
        return new PBRegistry<>(classifiers);
    }

    @SafeVarargs
    public PBRegistry(Class<? extends Enum<?>>... classifiersIn) {
        this(Arrays.asList(classifiersIn));
    }

    @SuppressWarnings("unchecked")
    public PBRegistry(List<Class<? extends Enum<?>>> classifiersIn) {
        classifiers = ImmutableSet.<Class<? extends Enum<?>>>builder() //
                .addAll(classifiersIn) //
                .build();
        classes = new HashMap<>();
        val builder = ImmutableMap.<Class<? extends Enum<?>>, Map<Enum<?>, Set<T>>>builder();
        for (var clazz : classifiersIn) {
            try {
                builder.put(clazz, EnumMap.class.getConstructor(Class.class) //
                        .newInstance(clazz));
            } catch (InstantiationException | IllegalAccessException | IllegalArgumentException
                    | InvocationTargetException | NoSuchMethodException | SecurityException ignored) {
            }
        }
        categorized = builder.build();
        categorized.forEach((k, v) -> {
            for (var enumValue : k.getEnumConstants()) {
                v.put(enumValue, new HashSet<>());
            }
        });
    }

    public @Nullable T register(ResourceLocation key, T value) {
        return register(key.toString(), value);
    }

    public @Nullable T register(String key, T value) {
        val previousValue = classes.put(key, value);
        for (var clazz : classifiers) {
            var type = clazz.cast(ClassUtil.getFieldValueByClass(value, clazz));
            if (type == null) {
                continue;
            }

            var map = categorized.get(clazz);
            map.get(type).add(value);
        }
        return previousValue;
    }

    public T unregister(ResourceLocation key) {
        return unregister(key.toString());
    }

    public T unregister(String key) {
        val value = classes.remove(key);
        for (var clazz : classifiers) {
            var type = clazz.cast(ClassUtil.getFieldValueByClass(value, clazz));
            if (type == null) {
                continue;
            }

            var map = categorized.get(clazz);
            var set = map.get(type);
            set.remove(value);
        }
        return value;
    }

    public T get(ResourceLocation key) {
        return get(key.toString());
    }

    public T get(String key) {
        return classes.get(key);
    }

    @SuppressWarnings("unchecked")
    public <C extends Enum<C>> T getRandom(C key) {
        val set = getSet(key);
        return ((T[]) set.toArray())[new Random().nextInt(set.size())];
    }

    public <C extends Enum<C>> Set<T> getSet(C key) {
        return ImmutableSet.<T>builder().addAll(categorized.get(key.getClass()).get(key)).build();
    }

    public ImmutableMap<String, T> entries() {
        return ImmutableMap.<String, T>builder().putAll(classes).build();
    }

    public ImmutableSet<String> keySet() {
        return ImmutableSet.<String>builder().addAll(classes.keySet()).build();
    }

    public ImmutableSet<T> values() {
        return ImmutableSet.<T>builder().addAll(classes.values()).build();
    }

    public void forEach(BiConsumer<? super String, ? super T> action) {
        classes.forEach(action);
    }

    public static <T> PBRegistryBuilder<T> builder() {
        return new PBRegistryBuilder<T>();
    }
}

File: src/main/kotlin/com/zippeykeys/praisebe/registry/PBRegistry.kt
package com.zippeykeys.praisebe.registry

import com.google.common.collect.ImmutableMap
import com.google.common.collect.ImmutableSet
import com.zippeykeys.praisebe.util.getFieldValueByClass
import net.minecraft.util.ResourceLocation
import java.lang.reflect.InvocationTargetException
import java.util.Arrays
import java.util.HashMap
import java.util.HashSet
import java.util.function.BiConsumer

open class PBRegistry<T>(classifiersIn: List<Class<out Enum<*>>>) {
    protected val classifiers: ImmutableSet<Class<out Enum<*>>> = ImmutableSet.builder<Class<out Enum<*>>>()
            .addAll(classifiersIn)
            .build()

    protected val classes: MutableMap<String, T>

    protected val categorized: Map<Class<out Enum<*>>, MutableMap<out Enum<*>, MutableSet<T>>>

    @SafeVarargs
    constructor(vararg classifiersIn: Class<out Enum<*>>) : this(Arrays.asList<Class<out Enum<*>>>(*classifiersIn))

    init {
        classes = HashMap()

        val builder = ImmutableMap.builder<Class<out Enum<*>>, MutableMap<Enum<*>, MutableSet<T>>>()
        for (clazz in classifiersIn) {
            try {
                val map = mutableMapOf<Enum<*>, MutableSet<T>>()
                for (enumVal in clazz.enumConstants) {
                    map[enumVal] = HashSet()
                }
                builder.put(clazz, map)
            } catch (ignored: InstantiationException) {
            } catch (ignored: IllegalAccessException) {
            } catch (ignored: IllegalArgumentException) {
            } catch (ignored: InvocationTargetException) {
            } catch (ignored: NoSuchMethodException) {
            } catch (ignored: SecurityException) {
            }

        }
        categorized = builder.build()
        categorized.forEach { k, v ->
            for (enumValue in k.enumConstants) {
                v[enumValue] = HashSet()
            }
        }
    }

    operator fun set(key: ResourceLocation, value: T): T? {
        return set(key.toString(), value)
    }

    operator fun set(key: String, value: T): T? {
        val previousValue = classes.put(key, value)
        for (clazz in classifiers) {
            val type = clazz.cast(getFieldValueByClass(value!!, clazz)) ?: continue
            val map = categorized[clazz]
            map!![type]?.plus(value)
        }
        return previousValue
    }

    fun unregister(key: ResourceLocation): T? {
        return unregister(key.toString())
    }

    fun unregister(key: String): T? {
        val value = classes.remove(key)
        if (value != null)
            for (clazz in classifiers) {
                categorized[clazz]!![clazz.cast(getFieldValueByClass(value, clazz)) ?: continue]!!.remove(value)
            }
        return value
    }

    operator fun get(key: ResourceLocation): T? {
        return get(key.toString())
    }

    operator fun get(key: String): T? {
        return classes[key]
    }

    fun <C : Enum<C>> getRandom(key: C): T {
        return getSet(key).random()
    }

    fun <C : Enum<C>> getSet(key: C): Set<T> {
        return ImmutableSet.builder<T>()
                .addAll(categorized[key.javaClass]!![key]!!)
                .build()
    }

    operator fun contains(key: ResourceLocation): Boolean {
        return contains(key.toString())
    }

    operator fun contains(key: String): Boolean {
        return get(key) != null
    }

    fun entries(): ImmutableMap<String, T> {
        return ImmutableMap.builder<String, T>()
                .putAll(classes)
                .build()
    }

    fun keySet(): ImmutableSet<String> {
        return ImmutableSet.builder<String>()
                .addAll(classes.keys)
                .build()
    }

    fun values(): ImmutableSet<T> {
        return ImmutableSet.builder<T>()
                .addAll(classes.values)
                .build()
    }

    fun forEach(action: BiConsumer<in String, in T>) {
        classes.forEach(action)
    }
}